### PR TITLE
[Merged] Feature/firefox

### LINKED ISSUE #97: setup-mobile page crashed
Problem code:

```
// useActivatedUI.ts
export function useLastRecognizedIdentity() {
    return useValueRef(getActivatedUI().lastRecognizedIdentity)
}
```

There is no UI setup in option page since `getActivatedUI()` will yield `undefined`.

### PR DESCRIPTION
**Pull request by guanbinrui**
Support Firefox #25 

+ [x] Fix #97
+ [x] Fix auto pasting not working in FF;
+ [x] Fix request camera permission not working in FF;
+ [x] Fix double error messages when camera permission was denied;
+ [x] Fix user discover not working in FF;
+ [x] Fix page flash while loading QR code image;
+ [x] Support scanning QR code in web worker;

### CODE REVIEW

**Comment by Jack-Works** [comment]
Merge branch 'master' of github.com:/...

You can rebase master instead of leaving so many commits

**Comment by guanbinrui** [comment]
@Jack-Works it clean now.

**Comment by Jack-Works** [comment]
Not sure if this got fixed.
- [ ] QR scanning is not working
- [ ] Background page throws "Can't parse the page." it caused by User-Agent, we may fall back to a classic way to fetch the key (Note: feature/multiple-network has done this.)

**Comment by Misaka-0x447f** [comment]
Misaka online. testing qr-code.

**Comment by Misaka-0x447f** [comment]
firefox does not support request camera using this api. working on find a polyfill.

**Comment by guanbinrui** [comment]
I'm working on figuring out the difference between Chrome & FF when fetching other user's public key and try to cherry-pick the classic way from feature/multiple-network branch.

**Comment by Jack-Works** [comment]
@guanbinrui you can leave that problem. After multiple network get merged and this branch get rebased, I can fix it

**Comment by guanbinrui** [comment]
@Jack-Works I think it's not the best way of plunking user bio/post content by the 'classic way'. It's easy to break.
What I found that in raw user profile page there is a code snippet:

```
<div id="intro_container_id"><!-- user bio box html --></div>
```

so I wonder we can parse the html snippet in FF too.

**Comment by Misaka-0x447f** [comment]
using a polyfill of permissions.query is required? no available polyfill at this time on the npm, only permissions.request available but broken.

File: src/social-network/facebook.com/fetch-bio.ts [hunk 1]
```diff
@@ -6,7 +6,9 @@ export async function fetchFacebookBio(who: PersonIdentifier) {
     const url = getProfilePageUrlAtFacebook(who)
     const doc = await parseFacebookStaticHTML(url)
     if (!doc) throw new Error("Can't parse the page")
-    const bio = doc.querySelector<HTMLDivElement>('#bio')
+    const bio =
+        doc.querySelector<HTMLDivElement>('#bio') ||
+        doc.querySelector<HTMLDivElement>('div > div > div:nth-child(2) > div:nth-child(2)')
```
Comment (guanbinrui) [inline src/social-network/facebook.com/fetch-bio.ts]: I created a sandbox project for figuring out this magic selector.
> https://github.com/guanbinrui/locate-bio-node

**Comment by Misaka-0x447f** [comment]
> 
> 
> Do you have your prettier enabled?

Emmmmm No force use of prettier in this project. Maybe need to chain it in the webpack and git commit hook.

**Comment by Jack-Works** [comment]
Does this pr ready for review?

**Comment by Misaka-0x447f** [comment]
no the qr component performance problem is not resolved

> 
> 
> Does this pr ready for review?

File: config-overrides.js [hunk 2]
```diff
@@ -14,8 +14,12 @@ module.exports = function override(/** @type{import("webpack").Configuration} */
         contentscript: path.join(__dirname, './src/content-script.ts'),
         backgroundservice: path.join(__dirname, './src/background-service.ts'),
         injectedscript: path.join(__dirname, './src/extension/injected-script/index.ts'),
+        qrcode: path.join(__dirname, './src/web-workers/QRCode.ts'),
     }
     if (env !== 'development') delete config.entry.devtools
+
+    // Let bundle compatiable with web worker
+    config.output.globalObject = 'this'
```
Comment (Jack-Works) [inline config-overrides.js]: I think we should use `globalThis` instead

File: src/social-network-provider/facebook.com/parse-html.ts [hunk 3]
```diff
@@ -8,8 +8,13 @@ export async function parseFacebookStaticHTML(url: RequestInfo) {
...
-    const code = codeDom.innerHTML.replace('<!--', '').replace('-->', '')
-    const doc2 = parser.parseFromString(code, 'text/html')
-    return doc2
+    const rootDom = doc1.body.querySelector('#root')
+    if (codeDom) {
+        return parser.parseFromString(codeDom.innerHTML.replace('<!--', '').replace('-->', ''), 'text/html')
+    }
+    // <code /> node is absent in old version profile page since use root node instead
+    if (rootDom) {
+        return parser.parseFromString(rootDom.innerHTML, 'text/html')
```
Comment (Jack-Works) [inline src/social-network-provider/facebook.com/parse-html.ts]: Why not just return rootDom, codeDom need a re-parse is due to `<!--` `-->` comment in the content?

File: src/components/Welcomes/QRScanner/ShapeDetectionPolyfill.ts [hunk 4]
```diff
@@ -1,9 +1,17 @@
-import jsQR from 'jsqr'
-import { isNull } from 'lodash-es'
 /// <reference path="./ShapeDetectionSpec.d.ts" />
+import { isNull } from 'lodash-es'
+
+const noop = () => {}
 
 class BarcodeDetectorPolyfill implements BarcodeDetector {
-    constructor() {}
+    private worker: Worker = new Worker(browser.runtime.getURL('js/qrcode.js'))
```
Comment (Jack-Works) [inline src/components/Welcomes/QRScanner/ShapeDetectionPolyfill.ts]: We have a `getUrl` in the project that works for both WebExt and storybook. Use browser.runtime.getURL directly will cause this code to fail in the storybook

**Comment by guanbinrui** [comment]
> need to test more... it seems that not entire post is used as a thing to decrypt the post, though unable to decrypt may related to GUN.
> 
> image

I'm working on this bug. It looks like maskbook fail to recognize current user.
#/setup-mobile page is crash on this branch too which is find on master branch.

**Pull request closed without merging**